File: NET/API/Treehopper/Spi.cs
namespace Treehopper
{
    using System.Threading.Tasks;

    /// <summary>
    /// An SPI interface
    /// </summary>
    public interface Spi
    {
        /// <summary>
        /// Enable or disable the SPI module.
        /// </summary>
        bool Enabled { get; set; }

        /// <summary>
        /// Send/receive data
        /// </summary>
        /// <param name="dataToWrite">a byte array of the data to send. The length of the transaction is determined by the length of this array.</param>
        /// <param name="chipSelect">The chip select pin, if any, to use during this transaction.</param>
        /// <param name="chipSelectMode">The chip select mode to use during this transaction (if a CS pin is selected)</param>
        /// <param name="speedMhz">The speed to perform this transaction at.</param>
        /// <param name="burstMode">Whether to use one of the burst modes</param>
        /// <param name="spiMode">The SPI mode to use during this transaction.</param>
        /// <returns>An awaitable byte array with the received data.</returns>
        Task<byte[]> SendReceive(
            byte[] dataToWrite, 
            SpiChipSelectPin chipSelect = null, 
            ChipSelectMode chipSelectMode = ChipSelectMode.SpiActiveLow, 
            double speedMhz = 1, 
            SpiBurstMode burstMode = SpiBurstMode.NoBurst, 
            SpiMode spiMode = SpiMode.Mode00
        );
    }
}

File: Java/api/treehopper/src/main/java/io/treehopper/interfaces/Spi.java
package io.treehopper.interfaces;

import io.treehopper.enums.BurstMode;
import io.treehopper.enums.ChipSelectMode;
import io.treehopper.enums.SpiMode;

/**
 * SPI interface
 */
public interface Spi {
    /**
     * Gets whether the SPI interface is enabled
     * @return true if the interface is enabled; false otherwise
     */
    boolean isEnabled();

    /**
     * Sets whether the SPI interface is enabled
     * @param enabled true to enable the interface; false to disable it
     */
    void setEnabled(boolean enabled);

    /**
     * Exchange data with an SPI slave
     * @param dataToWrite the data to write
     * @return The data read from the peripheral
     */
    public byte[] SendReceive(byte[] dataToWrite);

    /**
     * Exchange data with an SPI slave
     * @param dataToWrite the data to write
     * @param chipSelect the chip select pin to use
     * @param chipSelectMode the chip select mode to use
     * @return The data read from the peripheral
     */
    public byte[] SendReceive(byte[] dataToWrite, SpiChipSelectPin chipSelect, ChipSelectMode chipSelectMode);

    /**
     * Exchange data with an SPI slave
     * @param dataToWrite the data to write
     * @param chipSelect the chip select pin to use
     * @param chipSelectMode the chip select mode to use
     * @param speedMhz the speed, in MHz, to use
     * @return The data read from the peripheral
     */
    public byte[] SendReceive(byte[] dataToWrite, SpiChipSelectPin chipSelect, ChipSelectMode chipSelectMode, double speedMhz);

    /**
     * Exchange data with an SPI slave
     * @param dataToWrite the data to write
     * @param chipSelect the chip select pin to use
     * @param chipSelectMode the chip select mode to use
     * @param speedMhz the speed, in MHz, to use
     * @param burstMode the burst mode (if any) to use
     * @return The data read from the peripheral
     */
    public byte[] SendReceive(byte[] dataToWrite, SpiChipSelectPin chipSelect, ChipSelectMode chipSelectMode, double speedMhz, BurstMode burstMode);

    /**
     * Exchange data with an SPI slave
     * @param dataToWrite the data to write
     * @param chipSelect the chip select pin to use
     * @param chipSelectMode the chip select mode to use
     * @param speedMhz the speed, in MHz, to use
     * @param burstMode the burst mode (if any) to use
     * @param spiMode the SPI mode to use
     * @return The data read from the peripheral
     */
    public byte[] SendReceive(byte[] dataToWrite, SpiChipSelectPin chipSelect, ChipSelectMode chipSelectMode, double speedMhz, BurstMode burstMode, SpiMode spiMode);
}